sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala

```diff
@@ -469,9 +469,16 @@ object StructType extends AbstractDataType {
           case leftField @ StructField(leftName, leftType, leftNullable, _) =>
             rightMapped.get(leftName)
               .map { case rightField @ StructField(_, rightType, rightNullable, _) =>
-                leftField.copy(
-                  dataType = merge(leftType, rightType),
-                  nullable = leftNullable || rightNullable)
+                Try {
+                  merge(leftType, rightType)
+                } match {
+                  case Success(dataType) =>
+                    leftField.copy(
+                      dataType = dataType,
+                      nullable = leftNullable || rightNullable)
+                  case Failure(e) =>
+                    throw new SparkException(s"Failed to merge field '$leftName': " + e.getMessage)
```

Comment:
Could we throw an `AnalysisException` with both sides, `left` and `right`? Thanks!